README.md
## @netless/white-snapshot

Take a snapshot of a white-web-sdk scene.

Requires white-web-sdk &ge; 2.16.20.

### Usage

```js
import { snapshot } from "@netless/white-snapshot";
// room = await sdk.joinRoom(...)
snapshot(room).then(canvas => {
  document.body.append(canvas);
});
```

### Options

```ts
snapshot(room, {
  scenePath: "/init",
  padding: 5,
  background: 'transparent',
  crop: null,
  crossorigin: false,
  src2dataurl: undefined,
}): Promise<HTMLCanvasElement | null>;
```

Returns `null` if failed.

| Option      | Type                                       | Default       | Description                                                                                             |
| ----------- | ------------------------------------------ | ------------- | ------------------------------------------------------------------------------------------------------- |
| scenePath   | string                                     | current scene | Default is `displayer.state.sceneState.scenePath`.                                                      |
| padding     | number                                     | 5             | Pixels to the border of canvas.                                                                         |
| background  | string                                     |               | Background color.                                                                                       |
| crop        | Rectangle                                  | null          | Apply crop on the snapshot. Note that the snapshot includes padding.                                    |
| crossorigin | boolean                                    | false         | Apply hack to `document.createElement('img')` to enable crossorigin images.                             |
| src2dataurl | (src: string) &rArr; Promise&lt;string&gt; |               | Due to a [limitation][1] in `html2canvas`, we must convert images to dataurl before feeding them to it. |

[1]: https://github.com/niklasvh/html2canvas/issues/592#issuecomment-727540799

### Develop

```bash
pnpm dev
```

### [Changelog](./CHANGELOG.md)

## License

MIT @ [netless](https://github.com/netless-io/white-snapshot)


CHANGELOG.md
## Changelog

### 0.4.2

- Updated `Displayer` typing, for better understand what it does.

### 0.4.1

- Added `src2dataurl` option.

  Due to a [limitation][1] in `html2canvas`, we must convert images to dataurl before feeding them to it.
  You can change this option to feed your own image loader, for example:

  ```js
  import { src2dataurl } from "@netless/white-snapshot";
  // Force ignore cache.
  snapshot({ ...options, src2dataurl: src => src2dataurl(src, true) });
  ```

[1]: https://github.com/niklasvh/html2canvas/issues/592#issuecomment-727540799

### 0.4.0

- Added back `html2canvas`.

### 0.3.1

- Fixed `padding` not work.\
  It still has a little visual bug.

- Added Options:

  ```ts
  interface SnapshotOptions {
    /**
     * Background color.
     *
     * @default transparent
     */
    background?: string;
  }
  ```

### 0.3.0

- Removed html2canvas dependency.\
  Now it must requires white-web-sdk &ge; 2.16.20.

### 0.2.2

- Fixed `html2canvas` usage may get empty output.\
  Now it works the same as 0.2.0.
- Added Options:

  ```ts
  interface SnapshotOptions {
    /**
     * Print which scene.
     *
     * @default displayer.state.sceneState.scenePath
     */
    scenePath?: string;
  }
  ```

### 0.2.1

- Support white-web-sdk 2.16.20's native canvas output.
- Added options:

  ```ts
  interface SnapshotOptions {
    /**
     * Use `html2canvas` to print SVG directly.
     *
     * @default false
     */
    html2canvas?: boolean;
    /**
     * Apply hack to all `document.createElement('img')` to include crossorigin attribute.
     * This option requires image server settings and therefore is not enabled by default.
     *
     * @default false
     */
    crossorigin?: boolean;
  }
  ```

### 0.2.0

- Support `html2canvas` based snapshot. Options:

  ```ts
  interface SnapshotOptions {
    /**
     * @default 5 (px)
     */
    padding?: number;
    /**
     * Apply crop on the snapshot. Note that the snapshot includes padding.
     *
     * @default null
     */
    crop?: Record<"x" | "y" | "width" | "height", number> | null;
  }

  function snapshot(
    displayer: Displayer,
    { padding, crop: crop_ }?: SnapshotOptions
  ): Promise<HTMLCanvasElement | null>;
  ```


test.ts
import { DefaultHotKeys, RenderEngine, WhiteWebSdk } from "white-web-sdk";
import { snapshot } from "./index";

const sdk = new WhiteWebSdk({
  appIdentifier: "Gh0NyZub25jZT0x/Amcm9sZT0w",
  renderEngine: RenderEngine.Canvas,
});

sdk
  .joinRoom({
    uid: "uid-" + Math.random().toString(36).slice(2),
    uuid: "7285abc0b62911ecbd89bfc026a26890",
    roomToken:
      "NETLESSROOM_YWs9eTBJOWsxeC1IVVo4VGh0NyZub25jZT0xNjQ5MzA0OTY3Njg0MDAmcm9sZT0wJnNpZz04YjA2N2Q1ZjAwZjU1NTBjM2E3NjNkMWIzNjMxOGUyMGUzMWRmOTI3ZmFiYjNmOTkzYTk0NTYwOWEwYTM4N2M3JnV1aWQ9NzI4NWFiYzBiNjI5MTFlY2JkODliZmMwMjZhMjY4OTA",
    hotKeys: {
      ...DefaultHotKeys,
      changeToEraser: "e",
      changeToPencil: "p",
    },
  })
  .then((room) => {
    (window as any).room = room;
    (window as any).snapshot = snapshot;

    room.bindHtmlElement(document.getElementById("whiteboard") as HTMLDivElement);
  });


server.ts
import http from "http";
import https from "https";

// Simple server to serve CORS image.
var img_orginal =
  "https://flat-storage.oss-accelerate.aliyuncs.com/cloud-storage/2022-02/21/59455e11-d5fe-4f88-9d50-39b824a5c94b/59455e11-d5fe-4f88-9d50-39b824a5c94b.jpeg";

var fetch_img = new Promise((resolve) => {
  https.get(img_orginal, (res) => {
    var chunks = [];
    res.on("data", (c) => chunks.push(c));
    res.on("end", () => {
      resolve(Buffer.concat(chunks));
    });
  });
});

var server = http.createServer((req, res) => {
  if (req.url === "/img.jpeg") {
    res.writeHead(200, {
      "Content-Type": "image/jpeg",
      "Access-Control-Allow-Origin": "http://localhost:3000",
    });
    fetch_img.then((buffer) => {
      // https://stackoverflow.com/questions/34391134/send-a-binary-buffer-to-client-through-http-serverresponse-in-node-js
      res.write(buffer, "binary");
      res.end(null, "binary");
    });
  } else {
    res.writeHead(404);
    res.end("not found");
  }
});

server.listen(4000, () => {
  console.log("serving http://localhost:4000/img.jpeg");
});


index.ts
import html2canvas from "html2canvas";

/**
 * Search the displaying svg element.
 * @param container The param you used in `room.bindHtmlElement`.
 * @returns The svg element or null if not found.
 */
function search_svg(container: HTMLDivElement) {
  return container.querySelector(".background~svg") as SVGElement;
}

function extract_size_from_svg(svg: SVGElement, width: number, height: number) {
  const viewBox = svg.getAttribute("viewBox");
  if (viewBox) {
    const view = viewBox.split(" ").map((e) => Number(e));
    [width, height] = view.slice(2);
  }
  return { width, height };
}

function next_frame() {
  return new Promise((resolve) => {
    (window.requestAnimationFrame || setTimeout)(resolve);
  });
}

async function next_frames(n: number) {
  while (n-- > 0) await next_frame();
}

function noop() {}

function call_fn<T>(fn: () => T) {
  return fn();
}

export interface SnapshotOptions {
  /**
   * Print which scene.
   *
   * @default displayer.state.sceneState.scenePath
   */
  scenePath?: string;

  /**
   * @default 5 (px)
   */
  padding?: number;

  /**
   * Background color.
   *
   * @default transparent
   */
  background?: string;

  /**
   * Apply crop on the snapshot. Note that the snapshot includes padding.
   *
   * @default null
   */
  crop?: Record<"x" | "y" | "width" | "height", number> | null;

  /**
   * Apply hack to all `document.createElement('img')` to include crossorigin attribute.
   * This option requires image server settings and therefore is not enabled by default.
   *
   * @default false
   */
  crossorigin?: boolean;

  /**
   * Due to a [limitation](https://github.com/niklasvh/html2canvas/issues/592#issuecomment-727540799)
   * in `html2canvas`, we must convert images to dataurl before feeding them to it.
   * You can set this option to replace the default image loader.
   *
   * You can obtain the default loader through:
   *
   * ```js
   * import { src2dataurl } from "@netless/white-snapshot"
   * ```
   */
  src2dataurl?: (src: string) => Promise<string>;
}

function wrapper_element({ width = 100, height = 100, padding = 0 } = {}) {
  const wrapper = document.createElement("div");
  Object.assign(wrapper.style, {
    position: "fixed",
    top: `-9999px`,
    left: `-9999px`,
    width: `${width}px`,
    height: `${height}px`,
    padding: `${padding}px`,
  });
  return wrapper;
}

export interface Displayer {
  state: {
    sceneState: {
      scenePath: string;
    };
    cameraState: {
      width: number;
      height: number;
    };
  };
  fillSceneSnapshot(scenePath: string, div: HTMLElement, width: number, height: number): void;
}

/**
 * Take a snapshot of a whiteboard scene.
 * @param displayer The `room` returned by `sdk.joinRoom()`.
 * @returns Promise of `null` if failed to render, or a rendered `canvas` element.
 */
export async function snapshot(
  displayer: Displayer,
  {
    scenePath: scenePath_,
    padding = 5,
    background,
    crop: crop_ = null,
    src2dataurl: src2dataurl_,
    crossorigin = false,
  }: SnapshotOptions = {}
) {
  const scenePath = scenePath_ || displayer.state.sceneState.scenePath;
  let { width, height } = displayer.state.cameraState;

  let wrapper = wrapper_element();
  document.body.appendChild(wrapper);

  const invoke = crossorigin ? hack_create_image_with_cross_origin : call_fn;

  // 1. Get real size from svg element.
  await invoke(async () => {
    displayer.fillSceneSnapshot(scenePath, wrapper, width, height);
    await next_frames(10);
  });
  const svg = search_svg(wrapper);
  if (!svg) {
    document.body.removeChild(wrapper);
    return null;
  }
  ({ width, height } = extract_size_from_svg(svg, width, height));

  // 2. Render canvas again with correct size,
  //    otherwise the dom elements won't be in the right place.
  document.body.removeChild(wrapper);
  wrapper = wrapper_element({ width, height, padding });
  document.body.appendChild(wrapper);
  await invoke(async () => {
    displayer.fillSceneSnapshot(scenePath, wrapper, width, height);
    await next_frames(10);
  });

  try {
    let canvas = await html2canvas(wrapper, {
      useCORS: true,
      backgroundColor: background || null,
      async onclone(doc: Document): Promise<void> {
        const images = Array.from(doc.getElementsByTagName("image"));
        const tasks: Promise<string>[] = [];
        const transformer = src2dataurl_ || src2dataurl;
        for (const img of images) {
          tasks.push(transformer(img.href.baseVal));
        }
        for (const img of images) {
          const dataurl = await tasks.shift();
          if (dataurl) img.href.baseVal = dataurl;
        }
      },
    });

    if (crop_) {
      canvas = crop(canvas, crop_);
    }
    return canvas;
  } catch (error) {
    console.error(error);
    return null;
  } finally {
    document.body.removeChild(wrapper);
  }
}

/**
 * Utility to cut a specific rect from a canvas.
 * @param canvas The canvas element.
 * @param rect The rect to cut.
 * @returns A new canvas element.
 */
export function crop(canvas: HTMLCanvasElement, rect: Record<"x" | "y" | "width" | "height", number>) {
  const newCanvas = document.createElement("canvas");
  const ctx = newCanvas.getContext("2d");
  if (!ctx) {
    throw new Error("Failed to create canvas context.");
  }
  newCanvas.width = rect.width;
  newCanvas.height = rect.height;
  newCanvas.style.cssText = canvas.style.cssText;
  ctx.drawImage(canvas, rect.x, rect.y, rect.width, rect.height, 0, 0, rect.width, rect.height);
  return newCanvas;
}

export function fill_background(canvas: HTMLCanvasElement, background: string) {
  const newCanvas = document.createElement("canvas");
  const ctx = newCanvas.getContext("2d");
  if (!ctx) {
    throw new Error("Failed to create canvas context.");
  }
  newCanvas.width = canvas.width;
  newCanvas.height = canvas.height;
  newCanvas.style.cssText = canvas.style.cssText;
  ctx.fillStyle = background;
  ctx.fillRect(0, 0, canvas.width, canvas.height);
  ctx.drawImage(canvas, 0, 0);
  return newCanvas;
}

export function add_padding(canvas: HTMLCanvasElement, padding: number) {
  const newCanvas = document.createElement("canvas");
  const ctx = newCanvas.getContext("2d");
  if (!ctx) {
    throw new Error("Failed to create canvas context.");
  }
  newCanvas.width = canvas.width + padding * 2;
  newCanvas.height = canvas.height + padding * 2;
  newCanvas.style.cssText = canvas.style.cssText;
  ctx.drawImage(canvas, padding, padding);
  return newCanvas;
}

let hacked = false; // helper to prevent recursive hack
export async function hack_create_image_with_cross_origin(cb: () => Promise<void>) {
  if (hacked) return await cb();

  hacked = true;
  const _createElement = document.createElement;
  (document as any).createElement = function (...args: any[]) {
    const result = _createElement.apply(this, args as any);
    if (result instanceof HTMLImageElement) {
      result.setAttribute("crossorigin", "anonymous");
    }
    return result;
  };

  try {
    return await cb();
  } finally {
    (document as any).createElement = _createElement;
    hacked = false;
  }
}

export async function src2dataurl(src: string, force = false) {
  const r = await fetch(src, { cache: force ? "no-store" : "default" });
  const blob = await r.blob();
  return new Promise<string>((resolve) => {
    const reader = new FileReader();
    reader.onload = () => resolve(reader.result as string);
    reader.readAsDataURL(blob);
  });
}
